pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java

```diff
@@ -439,6 +439,18 @@ private boolean validate(TableConfig tableConfig, String taskType) {
     return true;
   }
 
+  /**
+   * Get the max valid merge bucket end time for the segment, LONG.MIN_VALUE if the segment is not ready to merge
+   */
+  private long getMaxValidBucketEndTimeMsForSegment(SegmentZKMetadata segmentZKMetadata, long bucketMs, long bufferMs) {
+    if (segmentZKMetadata.getStartTimeMs() / bucketMs * bucketMs + bucketMs > System.currentTimeMillis() - bufferMs) {
```

Comment:
Could you add more description on how the max valid bucket end time is calculated?

Reply:
Thanks for reviewing, added some comments for the logic.

Reply:
I would separate out this into 2 lines for better readability.

```
e.g.
long bucketEndTimeForSegment = segmentZKMetadata.getStartTimeMs() / bucketMs * bucketMs + bucketMs
if (bucketEndTimeForSegment > System.currentTimeMillis() - bufferMs()) {
  return Long.MIN_VALUE
}
````


By the way, why do we need to use `startTimeMs` here? 

Things that we need to brainstorm: do we have a case where 
`(segmentZKMetadata.getEndTimeMs() / bucketMs + 1)` is different from `segmentZKMetadata.getStartTimeMs() / bucketMs * bucketMs + bucketMs`?

Reply:
We need the `startTimeMs` here because as long as the `first bucket <= now - bufferMs`, it's ready to merge. These two values can be different. Say we have a segment [10/1, 10/3], the first bucketEndTime (`segmentZKMetadata.getStartTimeMs() / bucketMs * bucketMs + bucketMs`) will be `10/2 00:00`, the last bucketEndTime (`(segmentZKMetadata.getEndTimeMs() / bucketMs + 1) * bucketMs`) will be `10/4 00:00`. If `now - bufferMs` = `10/2 00:00`, the segment is ready to merge. For this segment the algorithm will use `10/2 00:00` to calculate the delay metrics.

Reply:
Separated out into 2 lines.

Reply:
I see. In that case, let's add the comment that our task scheduler uses `startTime` based window computation when deciding a segment is valid for merge.